### DISCUSSION TITLE
Include destination type to ResolutionContext

### DISCUSSION THREAD

**Discussion by iotalambda** [LuckyPennySoftware/AutoMapper#3592] [Ideas]
This is likely related to #3363 

When using base types as `TDestination` in `CreateMap`, e.g. `CreateMap(typeof(MyVeryGenericSource<>), typeof(object))`, there seems to be no way to know in `ConvertUsing` what derived destination type was used in `mapper.Map<MyDerivedDestination>(myVeryGenericSource)`. At the moment we cannot, for example, instantiate an object of the exact destination type in `ConvertUsing`.

## Example
```c#
CreateMap(typeof(MyLibrarysProperty<>), typeof(object))
    .ConvertUsing((source, destination, context) =>
    {
        var property = source as MyLibrarysProperty;

        if (property.HasComplexObjectValue)
        {
            // In this situation I would love to fallback to whatever mapper has been provided from
            // the complex object's type to the destination type, but there is no way to know the destination
            // type, because here destination is just null object :(
            return context.Mapper.Map(property.Value, property.Value.GetType(), destination.GetType());
        }

        return property.Value;
    });
```

## Proposal
Include the `DestinationType` to `ResolutionContext` *or somewhere where it would be accessible*. For example,
```c#
public class ResolutionContext
{
    public IMappingOperationOptions Options { get; }
    public IDictionary<string, object> Items { get; }
    public IRuntimeMapper Mapper { get; }
    public Dictionary<ContextCacheKey, object> InstanceCache { get; }
    public Type DestinationType { get; }
}
```

How does this sound?

**Comment by lbargaoanu** [comment]
https://github.com/AutoMapper/AutoMapper/blob/565c9634e2a0b487c9b25bca20f67a43e0ee06a3/src/UnitTests/Mappers/CustomMapperTests.cs#L97

**Reply by iotalambda** [reply]
@lbargaoanu Care to elaborate this? I don't see how this is related to the issue.
